### ISSUE TITLE
Set custom caption position with afterChange event

### ISSUE THREAD

**Issue by prohtex** [dimsemenov/PhotoSwipe#919]
I am using the afterChange event, currItem.w, and currItem.fitRatio to calculate a custom width and position for the caption div.

```
function capSet(lightBox) {
    $('div.pswp__caption__center').show();
    if (!lightBox.currItem.title) {
        debugMsg("*there's no caption!*");
        return;
        }
    debugMsg("window width = "+$(window).width()+", window height = "+$(window).height());
    if (lightBox.currItem.fitRatio < 1) {
        thiswidth = Math.floor(lightBox.currItem.w*lightBox.currItem.fitRatio);
        } else {
        debugMsg("*fitRatio greater than 1 (small image)*");
        thiswidth = lightBox.currItem.w;
        }
    capwidth = Math.floor(($(window).width()-thiswidth)/4);
    marginwidth = Math.floor(($(window).width()-thiswidth)/5);
    capvpos = Math.floor($(window).height()/4);
    debugMsg("slide width = currItem.w ("+lightBox.currItem.w+") * currItemfitRatio ("+lightBox.currItem.fitRatio.toFixed(2)+") = "+thiswidth);
    $('div.pswp__caption__center').show();
    if ($(window).width() - thiswidth > 600) {
        debugMsg("*there is room for a right-hand caption*");
        $('div.pswp__caption__center').addClass("capright");
        $('div.pswp__caption__center').css({    "width" : capwidth,
                                                "top" : capvpos,
                                                "right": marginwidth
                                                });
        } else {
        $('div.pswp__caption__center').removeClass("capright");
        $('div.pswp__caption__center').css("width",capwidth);
        $('div.pswp__caption__center').width(thiswidth);
        }
    }

function launchGallery(items, options) {
    var lightBox = new PhotoSwipe($('.pswp')[0], PhotoSwipeUI_Default, items, options);
    lightBox.init();
    if (lightBox.likelyTouchDevice) { return true; } // touch device, don't mess with the captions
    debugMsg("init<hr>");
    capSet(lightBox);
    lightBox.updateSize(true);
    lightBox.listen('preventDragEvent', function(e) {
        debugMsg("*drag event ("+e.type+")*");
        if ($('div.pswp__caption__center').hasClass("capright")) {
            if (e.type == "mousedown") {
                $('div.pswp__caption__center').hide();
                } else {
                setTimeout(function(){ $('div.pswp__caption__center').show(); },1000);
                }
            }
        })
    lightBox.listen('mouseUsed', function() { debugMsg("*mouseused*"); });
    lightBox.listen('afterChange', function() {
        debugMsg("<hr>");
        capSet(lightBox);
        lightBox.updateSize(true);
        });
    lightBox.listen('resize', function() {
        debugMsg("<b>resize called</b>");
        capSet(lightBox);
        });
    }
```

Here is my test site link. I have a debug div in the upper L corner to show window events:

http://www.thing.gallery

The positioning works correctly most of the time. Sometimes however, the caption is not positioned correctly until the window is resized. Below see two identical slides with different positions:

http://www.thing.gallery/#&gid=front&pid=12-the-dutch-proverbs
http://www.thing.gallery/#&gid=front&pid=13-the-dutch-proverbs

![screen shot 2015-07-28 at 12 28 22 am](https://cloud.githubusercontent.com/assets/12946938/8926051/8fe3948e-34c1-11e5-932f-2eae75599271.png)
![screen shot 2015-07-28 at 12 28 19 am](https://cloud.githubusercontent.com/assets/12946938/8926050/8fe2214e-34c1-11e5-8f84-ef4c5f741880.png)

Here we see that one slide is getting a fitRatio of .55, while the other is .52. Why?

The other issue I have here is that when the slides are dragged horizontally (left to right), if the caption is positioned right, it will appear over the image. So, I have to hide the caption with preventDragEvent. This causes unnecessary flickering when slides are changed because preventDragEvent fires very often. A better way would be to hide the caption when horizontal drag starts, but how do I access this event? I would put pswp.shout somewhere? Where?

Thank you again Dmitry for all your help.

**Comment by dimsemenov** [comment]
I have plans to implement NYtimes'ish caption positioning in some future version, and even tried to implement it during the initial development, but there were some issues regarding versatility of such design, and this hiding-during-drag didn't look as good as expected.

I can't assist you with it for now, if you don't understand how bottom margin is calculated, take a look in  source of default UI https://github.com/dimsemenov/PhotoSwipe/blob/master/src/js/ui/photoswipe-ui-default.js#L336

**Comment by prohtex** [comment]
The issue appears to be with the fitRatio - if you resize the window you can see the caption "jump" from bottom to right: http://www.thing.gallery/#&gid=front&pid=12-the-dutch-proverbs

**Comment by prohtex** [comment]
I understand this is beyond the scope of a general bug, and I really appreciate you looking at it. Nevermind about the hiding, I'll figure that one out. I just can't figure out how the fitRatio is being calculated. This is the issue in a nutshell:

afterChange event
fitRatio of .35
position caption
call updateSize to fix spacing
fitRatio of .37 ????????
caption positioned incorrectly

![screen shot 2015-07-28 at 2 43 34 am](https://cloud.githubusercontent.com/assets/12946938/8928356/c85b30ae-34d2-11e5-869c-97ee5d540061.png)

**Comment by prohtex** [comment]
Can you tell me how currItem.fitRatio is calculated?

**Comment by prohtex** [comment]
Here's the revised function that (mostly) resolves the "jumping caption" problem. Its not exactly elegant but it works.

In order to properly hide the caption on slide change, I added a handler for desktop horizontal pan slidechange event.

https://github.com/dimsemenov/PhotoSwipe/pull/923

	function capSet(lightBox) {
		$('div.pswp__caption__center').show();
		if (!lightBox.currItem.title) {
			debugMsg("*there's no caption!*");
			return;
			} else {
			caplength = lightBox.currItem.title.length;
			}
		if (caplength > 300) {
			threshold = 800;
			} else if (caplength > 200) {
			threshold = 600;
			} else {
			threshold = 500;
			}
		debugMsg("caption length = "+lightBox.currItem.title.length+", threshold = "+threshold);
		debugMsg("window width = "+$(window).width()+", window height = "+$(window).height());
		realwidth = $("img[src$='"+lightBox.currItem.src+"']").width();
		realheight = $("img[src$='"+lightBox.currItem.src+"']").height();
		gap = $(window).width() - realwidth;
		capwidth = Math.floor(gap/3.25);
		marginwidth = Math.floor(gap/2)+realwidth+40;
		capvpos = Math.floor($(window).height()/4);
		divwidth = $('div.pswp__caption__center div.text').outerWidth(true);
		divheight = $('div.pswp__caption__center div.text').outerHeight(true);
		debugMsg("realwidth = "+realwidth+", realheight = "+realheight);
		debugMsg("div width = "+divwidth+", div height = "+divheight);
		debugMsg("gap = "+gap);
		var $dummy = $('div.pswp__caption__center').clone().removeClass('pswp__caption__center').css('opacity',0).appendTo('body');
		ifbheight = $dummy.width(realwidth).outerHeight(true);
		ifrheight = $dummy.width(capwidth).outerHeight(true);
		$dummy.remove();
		debugMsg("ifbheight = "+ifbheight+", ifrheight = "+ifrheight);
		rforce = 0;
		if ($(window).height() - realheight == 88 && gap > threshold-(caplength/3)) {
			debugMsg("*image positioned without space for caption*");
			rforce = 1;
			}
		if (gap > threshold && gap-divheight < threshold || gap < threshold && gap+divheight > threshold) {
			debugMsg("***BORDERLINE***");
			}
		if (gap > threshold || rforce) {
			debugMsg("*there is room for a right-hand caption*");
			$('div.pswp__caption__center').addClass("capright");
			$('div.pswp__caption__center').css({	"width" : capwidth,
													"top" : capvpos,
													"left": marginwidth
													});
		} else {
			debugMsg("*bottom caption*");
			$('div.pswp__caption__center').removeClass("capright");
			$('div.pswp__caption__center').css("width",capwidth);
			$('div.pswp__caption__center').width(realwidth);
			}
		}

	function launchGallery(items, options) {
	
		var lightBox = new PhotoSwipe($('.pswp')[0], PhotoSwipeUI_Default, items, options);
		lightBox.init();
		if (lightBox.likelyTouchDevice) { return true; } // touch device, don't mess with the captions
		debugMsg("<hr>");
		capSet(lightBox);
		lightBox.updateSize(true);
		
		lightBox.listen('horizPanStarted', function() {
			debugMsg("*desktop horizontal pan started*");
			$('div.pswp__caption__center div.text').hide();
			});

		lightBox.listen('pointerUp', function() {
			setTimeout(function(){ $('div.pswp__caption__center div.text').show(); },1000);
			});
		
		lightBox.listen('afterChange', function() {
			//debugMsg("<hr>");
			debugMsg("",1);
			capSet(lightBox);
			lightBox.updateSize(true);
			});
			
		lightBox.listen('resize', function() {
			debugMsg("<hr>");
			capSet(lightBox);
			});
			
		}

**Comment by diniscorreia** [comment]
prohtex, that's brilliant!

dimsemenov, do you still plan to implement something like this?

**Comment by prohtex** [comment]
Yes, I'd love to hear from @dimsemenov as its been awhile and my implementation still has some quirks!

**Comment by diniscorreia** [comment]
@prohtex, by the way, I'm trying to use bits of your code - but on the first click on the gallery it never returns the real width of the image (if I exit PhotoSwipe and click the gallery again, it does return the size). Any ideas?
